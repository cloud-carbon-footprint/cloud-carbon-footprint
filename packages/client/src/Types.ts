/*
 * Â© 2021 Thoughtworks, Inc.
 */

import React, { Dispatch, ReactNode, SetStateAction } from 'react'
import {
  EstimationResult,
  RecommendationResult,
  ServiceData,
} from '@cloud-carbon-footprint/common'
import { Filters, FiltersDateRange } from './common/FilterBar/utils/Filters'
import { DropdownSelections } from './common/FilterBar/utils/FiltersUtil'

export interface ServiceResult<T> {
  data: T[]
  loading: boolean
  error: Error | null
}

export interface cloudEstPerDay {
  x: Date
  y: number
  usesAverageCPUConstant?: boolean
  kilowattHours?: number
  cost?: number
}

export interface DropdownOption {
  key: string
  name: string
  cloudProvider?: string
}

export interface AllFilterOptionMap {
  [type: string]: DropdownOption
}

export interface FilterOptions {
  [filterOption: string]: DropdownOption[]
}

export interface FilterResultResponse {
  accounts: DropdownOption[]
  services?: DropdownOption[]
  recommendationTypes?: DropdownOption[]
  regions?: DropdownOption[]
}

export interface FiltersConfig {
  timeframe?: number
  dateRange?: MaybeFiltersDateRange
  options: DropdownSelections
}

export type FilterBarProps = {
  filters: Filters
  setFilters: Dispatch<SetStateAction<Filters>>
  filterOptions: FilterOptions
  filteredData?: FilterResults
  setCo2eUnit?: Dispatch<SetStateAction<Co2eUnit>>
}

export interface Page<T> {
  data: T[]
  page: number
}

export interface PageEntry {
  x: string[]
  y: number
}

export type MaybeFiltersDateRange = FiltersDateRange | null

export type ApexChartProps = {
  data: EstimationResult[]
  dataType?: string
}

export type SidePanelProps = {
  drawerWidth: number
  title: string
  children: ReactNode
  defaultIsOpen?: boolean
  openOnChange?: RecommendationRow
  onClose?: () => void
}

export type FilterProps = {
  filters: Filters
  setFilters: Dispatch<SetStateAction<Filters>>
  options?: FilterOptions
}

export type FilterLabelMapping = { [type in DropdownFilterOptions]?: string }

export type UnknownTypesMapping = {
  [type in DropdownFilterOptions]?: UnknownTypes
}

export type FilterResults =
  | EstimationResult[]
  | RecommendationResult[]
  | EmissionsAndRecommendationResults

export type DateRange = {
  min: Date | null
  max: Date | null
}

export type Source = {
  href: string
  title: string
}

export type ComparisonItem = {
  icon: React.ReactNode
  total: number
  textOne: string
  textTwo: string
  source: Source
}

export type RecommendationRow = RecommendationResult & {
  id: number
  co2eUnit: Co2eUnit
}

export type EmissionsAndRecommendationResults = {
  recommendations: RecommendationResult[]
  emissions: ServiceData[]
}

export enum ChartDataTypes {
  REGION = 'region',
  SERVICE = 'service',
  ACCOUNT = 'account',
}

export enum UnknownTypes {
  UNKNOWN_REGION = 'Unknown Region',
  UNKNOWN_SERVICE = 'Unknown Service',
  UNKNOWN_ACCOUNT = 'Unknown Account',
  UNKNOWN_RECOMMENDATION_TYPE = 'Unknown Recommendation Type',
}

export enum DropdownFilterOptions {
  SERVICES = 'services',
  CLOUD_PROVIDERS = 'cloudProviders',
  ACCOUNTS = 'accounts',
  RECOMMENDATION_TYPES = 'recommendationTypes',
  REGIONS = 'regions',
}

export const barChartCustomColors: string[] = [
  '#73B500',
  '#00791E',
  '#D99200',
  '#DF5200',
  '#790000',
]

export const filterLabels: FilterLabelMapping = {
  [DropdownFilterOptions.ACCOUNTS]: 'Accounts',
  [DropdownFilterOptions.SERVICES]: 'Services',
  [DropdownFilterOptions.CLOUD_PROVIDERS]: 'Cloud Providers',
  [DropdownFilterOptions.REGIONS]: 'Regions',
  [DropdownFilterOptions.RECOMMENDATION_TYPES]: 'Recommendation Types',
}

export const unknownOptionTypes: UnknownTypesMapping = {
  [DropdownFilterOptions.ACCOUNTS]: UnknownTypes.UNKNOWN_ACCOUNT,
  [DropdownFilterOptions.SERVICES]: UnknownTypes.UNKNOWN_SERVICE,
  [DropdownFilterOptions.REGIONS]: UnknownTypes.UNKNOWN_REGION,
}

export enum Co2eUnit {
  Kilograms = 'Kilograms',
  MetricTonnes = 'MetricTonnes',
}
