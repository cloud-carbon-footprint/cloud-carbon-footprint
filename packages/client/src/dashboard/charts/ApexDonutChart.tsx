/*
 * Â© 2020 ThoughtWorks, Inc. All rights reserved.
 */

import React, { FunctionComponent } from 'react'
import { useTheme } from '@material-ui/core/styles'
import Chart from 'react-apexcharts'

import { getChartColors } from '../../themes'
import { sumCO2ByServiceOrRegion } from '../transformData'
import { ApexChartProps } from './common/ChartTypes'

export const ApexDonutChart: FunctionComponent<ApexChartProps> = ({ data, dataType }) => {
  const theme = useTheme()
  const chartColors = getChartColors(theme)

  const donutData = sumCO2ByServiceOrRegion(data, dataType)

  const donutDataEntries: { serviceOrRegion: string; c02Value: number }[] = Object.entries(donutData)
    .map((item) => ({
      serviceOrRegion: item[0],
      c02Value: item[1],
    }))
    .sort((higherC02, lowerCO2) => lowerCO2.c02Value - higherC02.c02Value)

  const sortedServicesOrRegions = donutDataEntries.map((entry) => entry.serviceOrRegion)
  const sortedCO2Emissions = donutDataEntries.map((entry) => entry.c02Value)

  const options = {
    chart: {
      background: theme.palette.background.paper,
    },
    colors: chartColors,
    dataLabels: {
      style: {
        fontSize: '1rem',
      },
      dropShadow: {
        enabled: true,
        top: 0,
        left: 0,
        blur: 3,
        color: theme.palette.grey[800],
        opacity: 0.5,
      },
    },
    labels: sortedServicesOrRegions,
    legend: {
      position: 'bottom',
      offsetY: -8,
    },
    series: sortedCO2Emissions,
    stroke: {
      colors: [theme.palette.background.default],
    },
    title: {
      offsetY: -8,
      style: {
        fontSize: '24px',
      },
    },
    height: '500px',
    theme: {
      mode: theme.palette.type,
    },
    tooltip: {
      fillSeriesColor: false,
      y: {
        formatter: function (value: number) {
          return `${value.toFixed(3)} metric tons CO2e`
        },
      },
    },
  }

  return <Chart options={options} series={options.series} type="donut" height={options.height} />
}
