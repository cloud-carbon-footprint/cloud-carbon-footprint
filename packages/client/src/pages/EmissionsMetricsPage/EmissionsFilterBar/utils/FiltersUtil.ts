/*
 * Â© 2021 Thoughtworks, Inc.
 */

import { DropdownFilterOptions, DropdownOption, FilterOptions } from 'Types'
import createOptionChooser from './options'

export type DropdownSelections = {
  [key in DropdownFilterOptions]: DropdownOption[]
}

export function handleDropdownSelections(
  filterType: DropdownFilterOptions,
  selections: DropdownOption[],
  oldSelections: DropdownSelections,
  filterOptions: FilterOptions,
): DropdownSelections {
  return createOptionChooser(
    filterType,
    selections,
    oldSelections,
    filterOptions,
  ).choose()
}

export function numSelectedLabel(
  length: number,
  totalLength: number,
  type = 'Services',
): string {
  const lengthWithoutAllOption = totalLength - 1
  if (length === totalLength) {
    return `${type}: ${lengthWithoutAllOption} of ${lengthWithoutAllOption}`
  } else {
    return `${type}: ${length} of ${lengthWithoutAllOption}`
  }
}
