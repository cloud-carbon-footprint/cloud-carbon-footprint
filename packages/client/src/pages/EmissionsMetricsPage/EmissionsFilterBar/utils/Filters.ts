/*
 * Â© 2021 Thoughtworks, Inc.
 */

import moment from 'moment'
import { EstimationResult } from '@cloud-carbon-footprint/common'
import * as FiltersUtil from './FiltersUtil'
import {
  ALL_ACCOUNTS_DROPDOWN_OPTION,
  ALL_SERVICES_DROPDOWN_OPTION,
  CLOUD_PROVIDER_OPTIONS,
} from './DropdownConstants'
import {
  FilterResultResponse,
  UnknownTypes,
  DropdownOption,
  FilterOptions,
  DropdownFilterOptions,
} from 'Types'

type MaybeDateRange = DateRange | null
type MaybeMoment = moment.Moment | null

interface FiltersConfig {
  timeframe: number
  dateRange: MaybeDateRange
  [DropdownFilterOptions.SERVICES]: DropdownOption[]
  [DropdownFilterOptions.CLOUD_PROVIDERS]: DropdownOption[]
  [DropdownFilterOptions.ACCOUNTS]: DropdownOption[]
}

const defaultFiltersConfig = {
  timeframe: 36,
  dateRange: null,
  [DropdownFilterOptions.SERVICES]: [ALL_SERVICES_DROPDOWN_OPTION],
  [DropdownFilterOptions.CLOUD_PROVIDERS]: CLOUD_PROVIDER_OPTIONS,
  [DropdownFilterOptions.ACCOUNTS]: [ALL_ACCOUNTS_DROPDOWN_OPTION],
}

export const filtersConfigGenerator = ({
  accounts,
  services,
}: FilterResultResponse): FiltersConfig => {
  const configValues = {
    accounts: [ALL_ACCOUNTS_DROPDOWN_OPTION, ...accounts],
    services: [ALL_SERVICES_DROPDOWN_OPTION, ...services],
  }

  return Object.assign(defaultFiltersConfig, configValues)
}

export class Filters {
  readonly timeframe: number
  readonly dateRange: MaybeDateRange
  readonly services: DropdownOption[]
  readonly cloudProviders: DropdownOption[]
  readonly accounts: DropdownOption[]

  constructor(config: FiltersConfig = defaultFiltersConfig) {
    this.timeframe = config.timeframe
    this.dateRange = config.dateRange
    this.services = config.services
    this.cloudProviders = config.cloudProviders
    this.accounts = config.accounts
  }

  withTimeFrame(timeframe: number): Filters {
    return this.timeframe === timeframe
      ? this
      : new Filters({ ...this, dateRange: null, timeframe })
  }

  withDropdownOption(
    dropdownOption: DropdownOption[],
    filterOptions: FilterOptions,
    dropdownFilter: DropdownFilterOptions,
  ): Filters {
    const oldSelections = {
      services: this.services,
      accounts: this.accounts,
      cloudProviders: this.cloudProviders,
    }
    return new Filters({
      ...this,
      ...FiltersUtil.handleDropdownSelections(
        dropdownFilter,
        dropdownOption,
        oldSelections,
        filterOptions,
      ),
    })
  }

  // withServices(
  //   services: DropdownOption[],
  //   filterOptions: FilterOptions,
  // ): Filters {
  //   const oldSelections = {
  //     services: this.services,
  //     accounts: this.accounts,
  //     cloudProviders: this.cloudProviders,
  //   }
  //   return new Filters({
  //     ...this,
  //     ...FiltersUtil.handleDropdownSelections(
  //       FiltersUtil.DropdownFilter.SERVICES,
  //       services,
  //       oldSelections,
  //       filterOptions,
  //     ),
  //   })
  // }
  //
  // withAccounts(
  //   accounts: DropdownOption[],
  //   filterOptions: FilterOptions,
  // ): Filters {
  //   const oldSelections = {
  //     services: this.services,
  //     accounts: this.accounts,
  //     cloudProviders: this.cloudProviders,
  //   }
  //   return new Filters({
  //     ...this,
  //     ...FiltersUtil.handleDropdownSelections(
  //       FiltersUtil.DropdownFilter.ACCOUNTS,
  //       accounts,
  //       oldSelections,
  //       filterOptions,
  //     ),
  //   })
  // }
  //
  // withCloudProviders(
  //   cloudProviders: DropdownOption[],
  //   filterOptions: FilterOptions,
  // ): Filters {
  //   const oldSelections = {
  //     services: this.services,
  //     accounts: this.accounts,
  //     cloudProviders: this.cloudProviders,
  //   }
  //   return new Filters({
  //     ...this,
  //     ...FiltersUtil.handleDropdownSelections(
  //       FiltersUtil.DropdownFilter.CLOUD_PROVIDERS,
  //       cloudProviders,
  //       oldSelections,
  //       filterOptions,
  //     ),
  //   })
  // }

  withDateRange(dateRange: DateRange): Filters {
    if (dateRange.isComplete()) {
      return new Filters({
        ...this,
        timeframe: -1,
        dateRange,
      })
    } else {
      return new Filters({
        ...this,
        dateRange,
      })
    }
  }

  serviceLabel(serviceOptions: DropdownOption[]): string {
    return FiltersUtil.numSelectedLabel(
      this.services.length,
      serviceOptions.length,
    )
  }

  cloudProviderLabel(): string {
    return FiltersUtil.numSelectedLabel(
      this.cloudProviders.length,
      CLOUD_PROVIDER_OPTIONS.length,
      'Cloud Providers',
    )
  }

  accountLabel(accountOptions: DropdownOption[]): string {
    return FiltersUtil.numSelectedLabel(
      this.accounts.length,
      accountOptions.length,
      'Accounts',
    )
  }

  filter(rawResults: EstimationResult[]): EstimationResult[] {
    const resultsFilteredByTime = this.getResultsFilteredByTime(rawResults)
    const resultsFilteredByService = this.getResultsFilteredByService(
      resultsFilteredByTime,
    )
    return this.getResultsFilteredByAccount(resultsFilteredByService)
  }

  private getResultsFilteredByAccount(
    resultsFilteredByService: EstimationResult[],
  ): EstimationResult[] {
    const allAccountsSelected = this.accounts.includes(
      ALL_ACCOUNTS_DROPDOWN_OPTION,
    )
    return resultsFilteredByService
      .map((estimationResult) => {
        const filteredServiceEstimates =
          estimationResult.serviceEstimates.filter((serviceEstimate) => {
            return (
              this.accounts.some(
                (account) =>
                  (account.name.includes(UnknownTypes.UNKNOWN_ACCOUNT) &&
                    serviceEstimate.accountName === null) ||
                  account.name === serviceEstimate.accountName,
              ) || allAccountsSelected
            )
          })
        return {
          timestamp: estimationResult.timestamp,
          serviceEstimates: filteredServiceEstimates,
        }
      })
      .filter(
        (estimationResult) => !!estimationResult?.serviceEstimates?.length,
      )
  }

  private getResultsFilteredByService(
    resultsFilteredByTime: EstimationResult[],
  ): EstimationResult[] {
    const allServicesSelected = this.services.includes(
      ALL_SERVICES_DROPDOWN_OPTION,
    )
    return resultsFilteredByTime.map((estimationResult) => {
      const filteredServiceEstimates = estimationResult.serviceEstimates.filter(
        (serviceEstimate) => {
          return (
            this.services.some(
              (service) =>
                (service.key.includes(UnknownTypes.UNKNOWN_SERVICE) &&
                  serviceEstimate.serviceName === null) ||
                service.key === serviceEstimate.serviceName,
            ) || allServicesSelected
          )
        },
      )
      return {
        timestamp: estimationResult.timestamp,
        serviceEstimates: filteredServiceEstimates,
      }
    })
  }

  private getResultsFilteredByTime(
    rawResults: EstimationResult[],
  ): EstimationResult[] {
    const today = moment.utc()
    let start: moment.Moment
    let end: moment.Moment
    if (this.timeframe < 0 && this.dateRange?.startDate) {
      start = this.dateRange.startDate
      end = this.dateRange.endDate || today
    } else {
      end = today
      start = end.clone().subtract(this.timeframe, 'M')
    }

    return rawResults.filter((estimationResult: EstimationResult) =>
      moment.utc(estimationResult.timestamp).isBetween(start, end, 'day', '[]'),
    )
  }
}

export class DateRange {
  readonly startDate: MaybeMoment
  readonly endDate: MaybeMoment

  constructor(startDate: MaybeMoment, endDate: MaybeMoment) {
    this.startDate = startDate
    this.endDate = endDate
  }

  isComplete(): boolean {
    return this.startDate !== null && this.endDate !== null
  }
}
