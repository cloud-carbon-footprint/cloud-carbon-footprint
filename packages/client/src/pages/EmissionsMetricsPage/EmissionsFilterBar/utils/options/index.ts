/*
 * Â© 2021 Thoughtworks, Inc.
 */

import { DropdownOption, FilterOptions, DropdownFilterOptions } from 'Types'
import { OptionChooser } from './OptionChooser'
import { DropdownSelections } from '../FiltersUtil'
import { CloudProviderChooser } from './CloudProviderChooser'
import { AccountChooser } from './AccountChooser'
import { ServiceChooser } from './ServiceChooser'

export default function createOptionChooser(
  filterType: DropdownFilterOptions,
  selections: DropdownOption[],
  oldSelections: DropdownSelections,
  filterOptions: FilterOptions,
): OptionChooser {
  switch (filterType) {
    case DropdownFilterOptions.CLOUD_PROVIDERS:
      return new CloudProviderChooser(selections, oldSelections, filterOptions)
    case DropdownFilterOptions.ACCOUNTS:
      return new AccountChooser(selections, oldSelections, filterOptions)
    case DropdownFilterOptions.SERVICES:
      return new ServiceChooser(selections, oldSelections, filterOptions)
  }
}
