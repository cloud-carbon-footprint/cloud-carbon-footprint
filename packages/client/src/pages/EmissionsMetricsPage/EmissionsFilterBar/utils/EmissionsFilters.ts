/*
 * Â© 2021 Thoughtworks, Inc.
 */

import moment from 'moment'
import { Filters } from '../../../../common/FilterBar/utils/Filters'
import {
  DropdownFilterOptions,
  DropdownOption,
  FilterOptions,
  FilterResultResponse,
  FiltersConfig,
  unknownOptionTypes,
} from '../../../../Types'
import {
  ALL_ACCOUNTS_DROPDOWN_OPTION,
  ALL_DROPDOWN_FILTER_OPTIONS,
  ALL_SERVICES_DROPDOWN_OPTION,
  CLOUD_PROVIDER_OPTIONS,
} from '../../../../common/FilterBar/utils/DropdownConstants'
import { EstimationResult, ServiceData } from '@cloud-carbon-footprint/common'
import { AccountChooser } from './options/AccountChooser'
import { DropdownSelections } from '../../../../common/FilterBar/utils/FiltersUtil'
import { OptionChooser } from '../../../../common/FilterBar/utils/OptionChooser'
import { CloudProviderChooser } from './options/CloudProviderChooser'
import { ServiceChooser } from './options/ServiceChooser'

const defaultConfig: FiltersConfig = {
  timeframe: 36,
  dateRange: null,
  options: {
    [DropdownFilterOptions.SERVICES]: [ALL_SERVICES_DROPDOWN_OPTION],
    [DropdownFilterOptions.CLOUD_PROVIDERS]: CLOUD_PROVIDER_OPTIONS,
    [DropdownFilterOptions.ACCOUNTS]: [ALL_ACCOUNTS_DROPDOWN_OPTION],
  },
}

export class EmissionsFilters extends Filters {
  constructor(config = defaultConfig) {
    super(config)
  }

  static generateConfig(newOptions: FilterResultResponse): FiltersConfig {
    return super.generateConfig(newOptions, defaultConfig)
  }

  create(config: FiltersConfig): Filters {
    return new EmissionsFilters(config)
  }

  createOptionChooser(
    filterType: DropdownFilterOptions,
    selections: DropdownOption[],
    oldSelections: DropdownSelections,
    filterOptions: FilterOptions,
  ): OptionChooser {
    switch (filterType) {
      case DropdownFilterOptions.CLOUD_PROVIDERS:
        return new CloudProviderChooser(
          selections,
          oldSelections,
          filterOptions,
        )
      case DropdownFilterOptions.ACCOUNTS:
        return new AccountChooser(selections, oldSelections, filterOptions)
      case DropdownFilterOptions.SERVICES:
        return new ServiceChooser(selections, oldSelections, filterOptions)
    }
  }

  filter(rawResults: EstimationResult[]): EstimationResult[] {
    const resultsFilteredByTime = this.getResultsFilteredByTime(rawResults)
    const resultsFilteredByService = this.getResultsFilteredBy(
      DropdownFilterOptions.SERVICES,
      resultsFilteredByTime,
    )
    return this.getResultsFilteredBy(
      DropdownFilterOptions.ACCOUNTS,
      resultsFilteredByService,
    ).filter((estimationResult) => !!estimationResult?.serviceEstimates?.length)
  }

  private getResultsFilteredByTime(
    previousResults: EstimationResult[],
  ): EstimationResult[] {
    const today = moment.utc()
    let start: moment.Moment
    let end: moment.Moment
    if (this.timeframe < 0 && this.dateRange?.startDate) {
      start = this.dateRange.startDate
      end = this.dateRange.endDate || today
    } else {
      end = today
      start = end.clone().subtract(this.timeframe, 'M')
    }

    return previousResults.filter((estimationResult: EstimationResult) =>
      moment.utc(estimationResult.timestamp).isBetween(start, end, 'day', '[]'),
    )
  }

  private getResultsFilteredBy(
    type: string,
    previousResults: EstimationResult[],
  ): EstimationResult[] {
    const hasAllOptionsSelected = this.options[type].includes(
      ALL_DROPDOWN_FILTER_OPTIONS[type],
    )
    return previousResults.map((estimationResult) => {
      const filteredServiceEstimates = estimationResult.serviceEstimates.filter(
        (serviceEstimate) =>
          this.options[type].some((dropdownOption) =>
            this.isUnknownOrSameName(dropdownOption, type, serviceEstimate),
          ) || hasAllOptionsSelected,
      )
      return {
        timestamp: estimationResult.timestamp,
        serviceEstimates: filteredServiceEstimates,
        periodStartDate: estimationResult.periodStartDate,
        periodEndDate: estimationResult.periodEndDate,
        groupBy: estimationResult.groupBy,
      }
    })
  }

  private isUnknownOrSameName(
    dropdownOption: DropdownOption,
    type: string,
    serviceEstimate: ServiceData,
  ) {
    const name =
      type === DropdownFilterOptions.ACCOUNTS ? 'accountName' : 'serviceName'
    return (
      (dropdownOption.name.includes(unknownOptionTypes[type]) &&
        serviceEstimate[name] === null) ||
      dropdownOption.name === serviceEstimate[name]
    )
  }
}
