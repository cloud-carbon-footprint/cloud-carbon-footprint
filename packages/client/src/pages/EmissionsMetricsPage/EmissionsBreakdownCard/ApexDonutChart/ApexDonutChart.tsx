/*
 * Â© 2021 Thoughtworks, Inc.
 */

import React, { FunctionComponent } from 'react'
import { useTheme } from '@material-ui/core/styles'
import Chart from 'react-apexcharts'

import { CCFTheme, getChartColors } from '../../../../utils/themes'
import { sumCO2ByServiceOrRegion } from '../../../../utils/helpers'
import { ApexChartProps } from '../../../../Types'

export const ApexDonutChart: FunctionComponent<ApexChartProps> = ({
  data,
  dataType,
}) => {
  const theme = useTheme() as CCFTheme
  const chartColors = getChartColors(theme)

  const donutData = sumCO2ByServiceOrRegion(data, dataType)

  const donutDataEntries: {
    serviceOrRegion: string
    c02Value: number
  }[] = Object.entries(donutData)
    .map((item) => ({
      serviceOrRegion: item[0],
      c02Value: item[1][1],
    }))
    .sort((higherC02, lowerCO2) => lowerCO2.c02Value - higherC02.c02Value)

  const sortedServicesOrRegions = donutDataEntries.map(
    (entry) => entry.serviceOrRegion,
  )
  const sortedCO2Emissions = donutDataEntries.map((entry) => entry.c02Value)

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const options: any = {
    chart: {
      background: theme.palette.background.paper,
    },
    colors: chartColors,
    dataLabels: {
      style: {
        fontSize: '1rem',
      },
      dropShadow: {
        enabled: true,
        top: 0,
        left: 0,
        blur: 3,
        color: theme.palette.grey[800],
        opacity: 0.5,
      },
    },
    labels: sortedServicesOrRegions,
    legend: {
      position: 'bottom',
      offsetY: -8,
    },
    series: sortedCO2Emissions,
    stroke: {
      colors: [theme.palette.background.default],
    },
    title: {
      offsetY: -8,
      style: {
        fontSize: '24px',
      },
    },
    height: '500px',
    theme: {
      mode: theme.palette.type,
    },
    tooltip: {
      fillSeriesColor: false,
      y: {
        formatter: function (value: number) {
          return `${value.toFixed(3)} metric tons`
        },
      },
    },
  }

  return (
    <Chart
      options={options}
      series={options.series}
      type="donut"
      height={options.height}
    />
  )
}
