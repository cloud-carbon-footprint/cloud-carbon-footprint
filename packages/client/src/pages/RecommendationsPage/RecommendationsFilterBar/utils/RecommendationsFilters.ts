/*
 * Â© 2021 Thoughtworks, Inc.
 */

import {
  RecommendationResult,
  ServiceData,
} from '@cloud-carbon-footprint/common'
import { Filters } from '../../../../common/FilterBar/utils/Filters'
import {
  DropdownFilterOptions,
  DropdownOption,
  EmissionsAndRecommendationResults,
  FilterOptions,
  FilterResultResponse,
  FiltersConfig,
  unknownOptionTypes,
} from '../../../../Types'
import {
  ALL_DROPDOWN_FILTER_OPTIONS,
  CLOUD_PROVIDER_OPTIONS,
} from '../../../../common/FilterBar/utils/DropdownConstants'
import { DropdownSelections } from '../../../../common/FilterBar/utils/FiltersUtil'
import { OptionChooser } from '../../../../common/FilterBar/utils/OptionChooser'
import { AccountChooser } from './options/AccountChooser'
import { CloudProviderChooser } from './options/CloudProviderChooser'
import { RegionChooser } from './options/RegionChooser'
import { RecommendationTypeChooser } from './options/RecommendationTypeChooser'

type ServiceDataOrRecommendationResult = (ServiceData | RecommendationResult)[]

const defaultConfig: FiltersConfig = {
  options: {
    [DropdownFilterOptions.ACCOUNTS]: [ALL_DROPDOWN_FILTER_OPTIONS.accounts],
    [DropdownFilterOptions.CLOUD_PROVIDERS]: CLOUD_PROVIDER_OPTIONS,
    [DropdownFilterOptions.REGIONS]: [ALL_DROPDOWN_FILTER_OPTIONS.regions],
    [DropdownFilterOptions.RECOMMENDATION_TYPES]: [
      ALL_DROPDOWN_FILTER_OPTIONS.recommendationTypes,
    ],
  },
}

export class RecommendationsFilters extends Filters {
  constructor(config = defaultConfig) {
    super(config)
  }

  static generateConfig(newOptions: FilterResultResponse): FiltersConfig {
    return super.generateConfig(newOptions, defaultConfig)
  }

  create(config: FiltersConfig): Filters {
    return new RecommendationsFilters(config)
  }

  createOptionChooser(
    filterType: DropdownFilterOptions,
    selections: DropdownOption[],
    oldSelections: DropdownSelections,
    filterOptions: FilterOptions,
  ): OptionChooser {
    switch (filterType) {
      case DropdownFilterOptions.ACCOUNTS:
        return new AccountChooser(selections, oldSelections, filterOptions)
      case DropdownFilterOptions.CLOUD_PROVIDERS:
        return new CloudProviderChooser(
          selections,
          oldSelections,
          filterOptions,
        )
      case DropdownFilterOptions.REGIONS:
        return new RegionChooser(selections, oldSelections, filterOptions)
      case DropdownFilterOptions.RECOMMENDATION_TYPES:
        return new RecommendationTypeChooser(
          selections,
          oldSelections,
          filterOptions,
        )
    }
  }

  filter(
    rawResults: EmissionsAndRecommendationResults,
  ): EmissionsAndRecommendationResults {
    const finalFilterResults = { ...rawResults }

    const resultTypes = Object.keys(rawResults)
    resultTypes.forEach((resultType) => {
      let filteredResult = rawResults[resultType]

      filteredResult = this.getResultsFilteredBy(
        DropdownFilterOptions.ACCOUNTS,
        filteredResult,
      )

      if (resultType === 'recommendations') {
        filteredResult = this.getResultsFilteredBy(
          DropdownFilterOptions.RECOMMENDATION_TYPES,
          filteredResult,
        )
      }

      finalFilterResults[resultType] = this.getResultsFilteredBy(
        DropdownFilterOptions.REGIONS,
        filteredResult,
      )
    })

    return finalFilterResults
  }

  private getResultsFilteredBy(
    type: string,
    previousResults: ServiceDataOrRecommendationResult,
  ): ServiceDataOrRecommendationResult {
    const hasAllOptionsSelected = this.options[type].includes(
      ALL_DROPDOWN_FILTER_OPTIONS[type],
    )

    return previousResults.filter(
      (result) =>
        this.options[type].some((dropdownOption) =>
          this.isUnknownOrSameName(dropdownOption, type, result),
        ) || hasAllOptionsSelected,
    )
  }

  private isUnknownOrSameName(
    dropdownOption: DropdownOption,
    type: string,
    result: ServiceData | RecommendationResult,
  ) {
    const typeFilterOptionMap = {
      [DropdownFilterOptions.ACCOUNTS]: 'accountName',
      [DropdownFilterOptions.RECOMMENDATION_TYPES]: 'recommendationType',
      [DropdownFilterOptions.REGIONS]: 'region',
    }

    const name = typeFilterOptionMap[type]

    return (
      (dropdownOption.name.includes(unknownOptionTypes[type]) &&
        result[name] === null) ||
      dropdownOption.name === result[name]
    )
  }
}
