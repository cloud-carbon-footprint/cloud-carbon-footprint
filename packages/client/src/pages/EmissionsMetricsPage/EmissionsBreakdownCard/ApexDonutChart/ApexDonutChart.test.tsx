/*
 * Â© 2021 Thoughtworks, Inc.
 */

import React from 'react'
import { create, ReactTestRenderer } from 'react-test-renderer'
import moment from 'moment'
import Chart from 'react-apexcharts'

import {
  EstimationResult,
  EmissionRatioResult,
  GroupBy,
} from '@cloud-carbon-footprint/common'

import { ServiceResult } from '../../../../Types'
import { ApexDonutChart } from './ApexDonutChart'
import { useRemoteEmissionService } from '../../../../utils/hooks'

jest.mock('apexcharts')
jest.mock('../../../../utils/themes')
jest.mock('../../../../utils/hooks/EmissionFactorServiceHook')

const mockedUseEmissionFactorService =
  useRemoteEmissionService as jest.MockedFunction<
    typeof useRemoteEmissionService
  >

const emissionsFactorData: EmissionRatioResult[] = [
  {
    region: 'us-west-1',
    mtPerKwHour: 0.000645,
    cloudProvider: 'AWS',
  },
  {
    region: 'us-west-2',
    mtPerKwHour: 0.000635,
    cloudProvider: 'AWS',
  },
  {
    region: 'us-west-3',
    mtPerKwHour: 0.000475,
    cloudProvider: 'AWS',
  },
  {
    region: 'us-west-4',
    mtPerKwHour: 0.000315,
    cloudProvider: 'AWS',
  },
]

describe('ApexDonutChart', () => {
  let fixture: ReactTestRenderer
  const timestamp: Date = moment('2019-08-10T00:00:00.000Z').toDate()
  const data: EstimationResult[] = [
    {
      timestamp: timestamp,
      periodStartDate: timestamp,
      periodEndDate: timestamp,
      groupBy: GroupBy.day,
      serviceEstimates: [
        {
          cloudProvider: 'AWS',
          accountId: 'some account id',
          accountName: 'some account',
          serviceName: 'ebs',
          kilowattHours: 0,
          co2e: 3000.014,
          cost: 0,
          region: 'us-west-2',
        },
        {
          cloudProvider: 'AWS',
          accountId: 'some account id',
          accountName: 'some account',
          serviceName: 's3',
          kilowattHours: 0,
          co2e: 1000.014,
          cost: 0,
          region: 'us-west-2',
        },
        {
          cloudProvider: 'AWS',
          accountId: 'some account id',
          accountName: 'some account',
          serviceName: 'ec2',
          kilowattHours: 0,
          co2e: 2000.014,
          cost: 0,
          region: 'us-west-2',
        },
        {
          cloudProvider: 'AWS',
          accountId: 'some account id',
          accountName: 'some account',
          serviceName: 'eks',
          kilowattHours: 0,
          co2e: 0.000014,
          cost: 0,
          region: 'us-west-2',
        },
      ],
    },
  ]
  beforeEach(() => {
    const mockReturnValue: ServiceResult<EmissionRatioResult> = {
      loading: false,
      data: emissionsFactorData,
    }
    mockedUseEmissionFactorService.mockReturnValue(mockReturnValue)
    fixture = create(<ApexDonutChart data={data} dataType="service" />)
  })

  afterEach(() => {
    fixture.unmount()
  })

  it('renders with correct configuration', () => {
    expect(fixture.toJSON()).toMatchSnapshot()
  })

  it('should pass sorted chart options to Chart component', () => {
    const chartOptionLabels =
      fixture.root.findByType(Chart).props?.options?.labels
    const chartOptionSeries =
      fixture.root.findByType(Chart).props?.options?.series

    expect(chartOptionLabels).toEqual(['ebs', 'ec2', 's3', 'eks'])
    expect(chartOptionSeries).toEqual([3000.014, 2000.014, 1000.014, 0.000014])
  })

  it('should format tool tip y values to 3 decimal places', () => {
    const yFormatter =
      fixture.root.findByType(Chart).props?.options?.tooltip?.y?.formatter
    expect(yFormatter).toBeDefined()
    expect(yFormatter(2000.0140003)).toEqual('2000.014 metric tons')
  })
})
