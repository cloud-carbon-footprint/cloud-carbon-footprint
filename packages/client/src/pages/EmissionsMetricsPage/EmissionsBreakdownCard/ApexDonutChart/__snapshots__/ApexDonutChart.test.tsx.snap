// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApexDonutChart renders with correct configuration 1`] = `<div />`;
